src/ray/stats/metric_exporter.h

```diff
@@ -100,38 +102,83 @@ class OpenCensusProtoExporter final : public opencensus::stats::StatsExporter::H
+        absl::make_unique<OpenCensusProtoExporter>(port,
+                                                   io_service,
+                                                   address,
+                                                   worker_id,
+                                                   report_batch_size,
+                                                   max_grpc_payload_size));
   }
 
   void ExportViewData(
       const std::vector<std::pair<opencensus::stats::ViewDescriptor,
                                   opencensus::stats::ViewData>> &data) override;
+  void SendData(const rpc::ReportOCMetricsRequest &request);
+
+  /// Processes data from the provided ViewData container by
+  ///   - Adding it into corresponding proto request payload
+  ///   - Submitting request payload to agent client (when it's reached target batch size
+  ///   or
+  ///     payload size limits)
+  ///
+  /// Upon returning of this method all of the time-series from the provided ViewData
+  /// container will either be a) contained inside provide proto request payload or b)
+  /// already submitted to the client
+  ///
+  /// \param view_descriptor, descriptor of the metric
+  /// \param view_data, data container aggregating time-series for this metric (across
+  /// different set of tags)
+  /// \param request_proto, target proto payload to embed metric values into
+  /// \param cur_batch_size current size of the batch (in terms of number of time-series
+  /// already added)
+  /// \param next_payload_size_check_at next batch size when payload (binary) size have
+  /// to be checked
+  void ProcessMetricsData(const opencensus::stats::ViewDescriptor &view_descriptor,
+                          const opencensus::stats::ViewData &view_data,
+                          rpc::ReportOCMetricsRequest &request_proto,
+                          size_t &cur_batch_size,
+                          size_t &next_payload_size_check_at);
+
+ protected:
+  rpc::ReportOCMetricsRequest createRequestProtoPayload();
+  size_t nextPayloadSizeCheckAt(size_t cur_batch_size);
+  bool handleBatchOverflows(const rpc::ReportOCMetricsRequest &request_proto,
+                            size_t cur_batch_size,
+                            size_t &next_batch_size_check_at);
+
   void addGlobalTagsToGrpcMetric(opencensus::proto::metrics::v1::Metric &metric);
-  void SendData(rpc::ReportOCMetricsRequest &request);
-  void UpdateMetricsData(const std::pair<opencensus::stats::ViewDescriptor,
-                                opencensus::stats::ViewData> &datum, rpc::ReportOCMetricsRequest &request_proto);
 
  private:
-  /// Call Manager for gRPC client.
-  rpc::ClientCallManager client_call_manager_;
   /// Lock to protect the client
   mutable absl::Mutex mu_;
   /// Client to call a metrics agent gRPC server.
-  std::unique_ptr<rpc::MetricsAgentClient> client_ ABSL_GUARDED_BY(&mu_);
+  std::shared_ptr<rpc::MetricsAgentClient> client_ ABSL_GUARDED_BY(&mu_);
```

Comment:
any reason why this should be a shared pointer? IIUC this is not used except within metrics_Exporter? (is it for testing?)

Reply:
Yes, this is needed for testing